src/Data/Conduit/Tar.hs

```diff
@@ -270,3 +245,482 @@ withEntries :: MonadThrow m
+                    pFileNameHeader <-
+                        packHeader $
+                        (defHeader offset)
+                        { headerFileNameSuffix = "././@LongLink"
+                        , headerPayloadSize = fPathLen
+                        , headerLinkIndicator = 76 -- 'L'
+                        , headerMagicVersion = gnuTarMagicVersion
+                        }
+                    yield pFileNameHeader
+                    yield fPath
+                    yield $ S.replicate (fromIntegral pad + 1) 0
+                    yield pHeader
+                    tarPayload 0 header tarFileInfo
+                Left exc -> throwM exc
+                Right header -> do
+                    packHeader header >>= yield
+                    tarPayload 0 header tarFileInfo
+        Nothing -> return offset
+
+
+
+
+
+-- | Create a tar archive by suppying `FileInfo`, which must be immediately
+-- followed by the file content, whenever its type is `FTNormal`.
+tar :: MonadResource m =>
+       ConduitM (Either FileInfo ByteString) ByteString m FileOffset
+tar = do
+    offset <- tarFileInfo 0
+    yield terminatorBlock
+    return $ offset + fromIntegral (S.length terminatorBlock)
+
+
+-- | Create a tar archive by suppying `Header`, which must be immediately
+-- followed by the file content, whenever its type is `FTNormal`.
+tarEntries :: MonadResource m =>
+            ConduitM (Either Header ByteString) ByteString m FileOffset
+tarEntries = do
+    offset <- tarHeader 0
+    yield terminatorBlock
+    return $ offset + fromIntegral (S.length terminatorBlock)
+
+
+
+-- | Turn a stream of file paths into a stream of `FileInfo` and file
+-- content. All paths will be decended into recursively.
+filePathConduit :: MonadResource m =>
+                   ConduitM FilePath (Either FileInfo ByteString) m ()
+filePathConduit = do
+    mfp <- await
+    case mfp of
+        Just fp -> do
+            fi <- liftIO $ getFileInfo $ S8.pack fp
+            case fileType fi of
+                    FTNormal         -> do
+                        yield (Left fi)
+                        sourceFile (S8.unpack (filePath fi)) .| mapC Right
+                    FTSymbolicLink _ ->
+                        yield (Left fi)
+                    FTDirectory      -> do
+                        yield (Left fi)
+                        sourceDirectory (S8.unpack (filePath fi)) .| filePathConduit
+                    fty              -> do
+                        leftover fp
+                        throwM $ TarCreationError $ "Unsupported file type: " ++ show fty
+            filePathConduit
+        Nothing -> return ()
+
+
+-- | Recursively tar all of the files and directories.
+tarFilePath :: MonadResource m => ConduitM FilePath ByteString m FileOffset
+tarFilePath = filePathConduit .| tar
+
+
+
+createTarball :: FilePath -- ^ File name for the tarball
+              -> [FilePath] -- ^ List of files and directories to include in the tarball
```

Comment:
Worth commenting on how generated file names are handled with regard to stripping the parent parent or not.

Reply:
Added some more documentation to couple other functions, which together hopefully clarify `createTarball` as well.